src/main/performanceMonitor.ts

```diff
@@ -0,0 +1,171 @@
+
+        ipcMain.on(EMIT_CONFIGURATION, this.handleConfigUpdate);
+    }
+
+    init = () => {
+        // Set that it's initted so that the powerMonitor functions correctly
+        this.isInitted = true;
+
+        // Run once because the first CPU value is always 0
+        this.runMetrics();
+
+        if (Config.enableMetrics) {
+            this.start();
+        }
+    };
+
+    registerView = (name: string, webContents: WebContents, serverId?: string) => {
+        log.debug('registerView', webContents.id, name);
+
+        webContents.on('did-finish-load', () => {
+            this.views.set(webContents.id, {name, webContents, serverId});
+        });
+    };
+
+    registerServerView = (name: string, webContents: WebContents, serverId: string) => {
+        log.debug('registerServerView', webContents.id, serverId);
+
+        webContents.on('did-finish-load', () => {
+            this.serverViews.set(webContents.id, {name, webContents, serverId});
+        });
+    };
+
+    unregisterView = (webContentsId: number) => {
+        log.debug('unregisterView', webContentsId);
+
+        this.views.delete(webContentsId);
+        this.serverViews.delete(webContentsId);
+    };
+
+    private start = () => {
+        if (!this.isInitted) {
+            return;
+        }
+
+        if (!Config.enableMetrics) {
+            return;
+        }
+
+        log.verbose('start');
+
+        if (this.updateInterval) {
+            clearInterval(this.updateInterval);
+        }
+        this.updateInterval = setInterval(this.sendMetrics, METRIC_SEND_INTERVAL);
+    };
+
+    private stop = () => {
+        log.verbose('stop');
+
+        clearInterval(this.updateInterval);
+        delete this.updateInterval;
+    };
+
+    private runMetrics = async () => {
+        const metricsMap: Map<string, Metrics> = new Map();
+
+        // Collect metrics for all of the Node processes
+        app.getAppMetrics().
+            filter((metric) => metric.type !== 'Tab').
+            forEach((metric) => {
+                metricsMap.set(metric.name ?? metric.type, {
+                    cpu: metric.cpu.percentCPUUsage,
+                    memory: metric.memory.privateBytes ?? metric.memory.workingSetSize,
+                });
+            });
+
+        const viewResolves: Map<number, () => void> = new Map();
+        const listener = (event: IpcMainEvent, name: string, metrics: Metrics) => {
+            metricsMap.set(name, metrics);
+            viewResolves.get(event.sender.id)?.();
+        };
+        ipcMain.on(METRICS_RECEIVE, listener);
+        const viewPromises = [...this.views.values(), ...this.serverViews.values()].map((view) => {
+            return new Promise<void>((resolve) => {
+                viewResolves.set(view.webContents.id, resolve);
+                view.webContents.send(METRICS_REQUEST, view.name, view.serverId);
+            });
+        });
+        await Promise.all(viewPromises);
```

Comment:
maybe here would be good to add some preventive code:
1) use `allSettled` instead of `all` so you can handle individual promise rejections (in case one of the promise rejects).
2) would be good to wrap this line and the following in a `try..finally`, so you can make sure that no matter what the removal of the listener will happen.
3) should there be a timeout maybe?

Reply:
i don't see a case where I wouldn't want the app to be loud about the `Promise` failing, so the `try/catch` doesn't seem as necessary. However I did want to add the timeout, thank you for reminding me :)